src/llvm-alloc-opt.cpp

```diff
@@ -0,0 +1,667 @@
+                    return false;
+                }
+            }
+            return true;
+        }
+        if (isa<AddrSpaceCastInst>(inst) || isa<BitCastInst>(inst)) {
+            push_inst(inst);
+            return true;
+        }
+        if (auto gep = dyn_cast<GetElementPtrInst>(inst)) {
+            APInt apoffset(sizeof(void*) * 8, cur.offset, true);
+            if (ignore_tag && (!gep->accumulateConstantOffset(*DL, apoffset) ||
+                               apoffset.isNegative()))
+                ignore_tag = false;
+            push_inst(inst);
+            cur.offset = apoffset.getLimitedValue();
+            // Check overflow
+            if (cur.offset == UINT64_MAX)
+                ignore_tag = false;
+            return true;
+        }
+        return false;
+    };
+
+    while (true) {
+        assert(cur.use_it != cur.use_end);
+        auto use = &*cur.use_it;
+        auto inst = dyn_cast<Instruction>(use->getUser());
+        ++cur.use_it;
+        if (!inst)
+            return false;
+        if (!check_inst(inst, use))
+            return false;
+        uses.insert(inst);
+        if (cur.use_it == cur.use_end) {
+            if (stack.empty())
+                return true;
+            cur = stack.back();
+            stack.pop_back();
+        }
+    }
+}
+
+// This function needs to handle all cases `AllocOpt::checkInst` can handle.
+// This function should not erase any safepoint so that the lifetime marker can find and cache
+// all the original safepoints.
+void AllocOpt::replaceUsesWith(Instruction *orig_inst, Instruction *new_inst,
+                               ReplaceUsesStack &stack)
+{
+    auto simple_replace = [&] (Instruction *orig_i, Instruction *new_i) {
+        if (orig_i->user_empty()) {
+            if (orig_i != orig_inst)
+                orig_i->eraseFromParent();
+            return true;
+        }
+        Type *orig_t = orig_i->getType();
+        Type *new_t = new_i->getType();
+        if (orig_t == new_t) {
+            orig_i->replaceAllUsesWith(new_i);
+            if (orig_i != orig_inst)
+                orig_i->eraseFromParent();
+            return true;
+        }
+        return false;
+    };
+    if (simple_replace(orig_inst, new_inst))
+        return;
+    assert(stack.empty());
+    ReplaceUsesFrame cur{orig_inst, new_inst};
+    auto finish_cur = [&] () {
+        assert(cur.orig_i->user_empty());
+        if (cur.orig_i != orig_inst) {
+            cur.orig_i->eraseFromParent();
+        }
+    };
+    auto push_frame = [&] (Instruction *orig_i, Instruction *new_i) {
+        if (simple_replace(orig_i, new_i))
+            return;
+        stack.push_back(cur);
+        cur = {orig_i, new_i};
+    };
+    // Both `orig_i` and `new_i` should be pointer of the same type
+    // but possibly different address spaces. `new_i` is always in addrspace 0.
+    auto replace_inst = [&] (Instruction *user) {
```

Comment:
You can technically mutate instructions in place, but I'm fine with this implementation as well.